zinc/src/test/scala/sbt/inc/BaseIncCompilerSpec.scala

```diff
@@ -0,0 +1,136 @@
+    true,
+    Array.empty
+  )
+
+  case class TestCompilerClasspathConfig(
+    override val classpath: Seq[File],
+    override val perClasspathEntryLookup: PerClasspathEntryLookup,
+    override val compiler: xsbti.compile.ScalaCompiler,
+    override val currentSetup: MiniSetup = mockedMiniSetup,
+    override val incOptions: IncOptions = IncOptionsUtil.defaultIncOptions()
+  ) extends CompilerClasspathConfig
+
+  protected def mockedConfig(mockedClasspath: Seq[File], mockedLookup: PerClasspathEntryLookup, tempDir: File): TestCompilerClasspathConfig =
+    TestCompilerClasspathConfig(mockedClasspath, mockedLookup, mockedCompiler(tempDir))
+}
+
+abstract class ClasspathEntry(in: DirectorySetup) {
+  def classpathEntry: File
+
+  def analysis: Option[Analysis]
+
+  in.register(this)
+}
+
+class VirtualDir(val name: String, val classes: Seq[String], val in: DirectorySetup) extends ClasspathEntry(in) {
+  val dir = new File(in.baseDir, name)
+  dir.mkdirs()
+
+  val classFiles = classes.map(classFile)
+  classFiles.foreach(_.createNewFile())
+
+  override def classpathEntry: File = dir
+
+  override def analysis: Option[Analysis] = None
+
+  def classFile(binaryName: String) = new File(dir, s"$binaryName.class")
+}
+
+case class VirtualJar(override val name: String, override val classes: Seq[String], override val in: DirectorySetup)
+  extends VirtualDir(name + "_tmp", classes, in) {
+  val location = new File(in.baseDir, s"$name.jar")
+
+  IO.zip(classFiles.map(f => f -> f.getName), location)
+
+  override def classpathEntry: File = location
+
+  override def analysis: Option[Analysis] = None
+}
+
+case class VirtualProject(name: String, in: DirectorySetup) extends ClasspathEntry(in) {
+  private var _analysis: Analysis = Analysis.Empty
+
+  def newSource(v: VirtualSource): Analysis = {
+    val newAnalysis = Analysis.empty(true)
+      .addSource(
+        v.sourceFile,
+        Seq(APIs.emptyAnalyzedClass.withName(v.name)),
+        Stamp.hash(v.sourceFile),
+        SourceInfos.emptyInfo,
+        Seq(NonLocalProduct(v.name, v.name, v.classFile, Stamp.lastModified(v.classFile))),
+        Nil, Nil, Nil, Nil
+      )
+    _analysis ++= newAnalysis
+    newAnalysis
+  }
+
+  val baseDir = new File(in.baseDir, name)
+  baseDir.mkdirs()
+
+  val src = new File(baseDir, "src")
+  src.mkdir()
+
+  val out = new File(baseDir, "out")
+  out.mkdir()
+
+  override def classpathEntry: File = out
+
+  override def analysis: Option[Analysis] = Some(_analysis)
+}
+
+case class VirtualSource(name: String, in: VirtualProject) {
+  val sourceFile = new File(in.src, s"$name.scala")
+  val classFile = new File(in.out, s"$name.class")
+
+  sourceFile.createNewFile()
+  classFile.createNewFile()
+
+  val analysis = in.newSource(this)
+}
+
+class DirectorySetup(val baseDir: File) {
+  private var myEntries = Seq.empty[ClasspathEntry]
+
+  def register(entry: ClasspathEntry) = myEntries :+= entry
+
+  def classpath = myEntries.map(_.classpathEntry)
```

Comment:
I think we should support any order of classpath - so unless we are not forcing classpath order we need to support each possibility